src/turbo-stream.spec.ts

```diff
@@ -572,3 +572,196 @@ test("should allow many nested promises without a memory leak", async () => {
   expect(currentDecoded.i).toBe(depth - 1);
   await decoded.done;
 });
+
+test.describe("ReadableStream", () => {
+  test("basic usage", async () => {
+    const input = new ReadableStream({
+      start(controller) {
+        controller.enqueue("foo");
+        controller.enqueue("bar");
+        setTimeout(() => {
+          controller.enqueue("foo");
+          controller.enqueue({ loba: "boba" });
+          controller.close();
+        }, 10);
+      },
+    });
+
+    const decoded = await decode(encode(input));
+
+    const inputReader = input.getReader();
+    const decodedReader = (decoded.value as typeof input).getReader();
+
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual({ value: "foo", done: false });
+    }
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual({ value: "bar", done: false });
+    }
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual({ value: "foo", done: false });
+    }
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual({ value: { loba: "boba" }, done: false });
+    }
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual({ value: undefined, done: true });
+    }
+    {
+      const inputValue = await inputReader.closed;
+      const decodedValue = await decodedReader.closed;
+      expect(inputValue).toEqual(decodedValue);
+      expect(inputValue).toEqual(undefined);
+    }
+    await decoded.done;
+  });
+
+  test("errors", async () => {
+    const input = new ReadableStream({
+      start(controller) {
+        controller.enqueue("foo");
+        controller.enqueue("bar");
+        setTimeout(() => {
+          controller.enqueue("foo");
+          controller.enqueue({ loba: "boba" });
+          setTimeout(() => {
+            controller.error(new Error("baz"));
+          }, 1);
+        }, 10);
+      },
+    });
+
+    const decoded = await decode(encode(input));
+
+    const inputReader = input.getReader();
+    const decodedReader = (decoded.value as typeof input).getReader();
+
+    {
+      const inputValue = await inputReader.read();
+      const decodedValue = await decodedReader.read();
```

Comment:
These tests currently test the "non-destructive" approach with `tee`, which makes it possible to still consume the original input stream.